internal/vault/credentials/aws.go

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package credentials
+
+import (
+	"context"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/hashicorp/go-secure-stdlib/awsutil"
+	secretsv1alpha1 "github.com/hashicorp/vault-secrets-operator/api/v1alpha1"
+	"github.com/hashicorp/vault-secrets-operator/internal/consts"
+)
+
+const (
+	AWSAnnotationRole            = "eks.amazonaws.com/role-arn"
+	AWSAnnotationAudience        = "eks.amazonaws.com/audience"
+	AWSAnnotationTokenExpiration = "eks.amazonaws.com/token-expiration"
+	AWSDefaultAudience           = "sts.amazonaws.com"
+	AWSDefaultTokenExpiration    = int64(86400)
+)
+
+var _ CredentialProvider = (*AWSCredentialProvider)(nil)
+
+type AWSCredentialProvider struct {
+	authObj           *secretsv1alpha1.VaultAuth
+	providerNamespace string
+	uid               types.UID
+}
+
+func (l *AWSCredentialProvider) GetNamespace() string {
+	return l.providerNamespace
+}
+
+func (l *AWSCredentialProvider) GetUID() types.UID {
+	return l.uid
+}
+
+func (l *AWSCredentialProvider) Init(ctx context.Context, client ctrlclient.Client, authObj *secretsv1alpha1.VaultAuth, providerNamespace string) error {
+	l.authObj = authObj
+	l.providerNamespace = providerNamespace
+
+	if l.authObj.Spec.AWS.AWSCredsRef != "" {
+		// If AWSCredsRef is not empty, get the secret and read the creds from
+		// there, use the secret UID as l.uid
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.AWSCredsRef,
+		}
+		credsSecret, err := getSecret(ctx, client, key)
+		if err != nil {
+			return err
+		}
+		l.uid = credsSecret.UID
```

Comment:
It's maybe not best practice, but I think there might be some scenarios where people have multiple VaultAuths pointing at the same static secret or service account, but using different Vault roles. In that case, the UID of those auth providers won't be unique. Should we generate our own UUID to avoid that?

Reply:
> It's maybe not best practice, but I think there might be some scenarios where people have multiple VaultAuths pointing at the same static secret or service account, but using different Vault roles. 

I think that should be fine, although as you stated maybe a little odd.

> In that case, the UID of those auth providers won't be unique. Should we generate our own UUID to avoid that?
I am not seeing the issue with that since in the end the Vault Client will be unique for each VaultAuth. Could you maybe unpack that a bit more.  What issues do you foresee, etc?

For the generation bit, the contract is that all credential providers should provide their own UUID (UID) that persists for the lifetime of that provider. We haven't encountered a situation where VSO has ever had to generate one on behalf of a credential provider.

Reply:
I'm not sure if it's a problem. But generally I would expect each distinct object to have a unique UID, so if there are multiple providers with the same UID, it feels like it could lead to surprising bugs. For example, UIDs are used to generate cache keys so UID collisions could in theory cause cache key collisions. And to be clear, I don't think that specific example _is_ a problem right now, my question is more generally "it looks like provider UID collisions can happen, are we ok with that?"

wrt "Should we generate our own UUID to avoid that?" - I don't really mind what solution we pick, but that was a possible suggestion we could copy what we do just below in the case that we don't have static or IRSA creds: https://github.com/hashicorp/vault-secrets-operator/blob/be159c35ceb97e477e3c244876688a960062e31c/internal/vault/credentials/aws.go#L73-L77

Reply:
Given the current design, we should definitely not be generating new credential provider UIDs.